File: Dsa-day244.cpp
#include <vector>
#include <iostream>

class Solution {
public:
    int numberOfSubmatrices(std::vector<std::vector<char>>& grid) {
        int rows = grid.size();
        int cols = grid[0].size();
        
        std::vector<std::vector<int>> prefixX(rows, std::vector<int>(cols, 0));
        std::vector<std::vector<int>> prefixY(rows, std::vector<int>(cols, 0));
        
        int count = 0;
        
        for (int i = 0; i < rows; ++i) {
            for (int j = 0; j < cols; ++j) {
                
                if (grid[i][j] == 'X') {
                    prefixX[i][j] = 1;
                }
                if (i > 0) {
                    prefixX[i][j] += prefixX[i-1][j];
                }
                if (j > 0) {
                    prefixX[i][j] += prefixX[i][j-1];
                }
                if (i > 0 && j > 0) {
                    prefixX[i][j] -= prefixX[i-1][j-1];
                }
                
                if (grid[i][j] == 'Y') {
                    prefixY[i][j] = 1;
                }
                if (i > 0) {
                    prefixY[i][j] += prefixY[i-1][j];
                }
                if (j > 0) {
                    prefixY[i][j] += prefixY[i][j-1];
                }
                if (i > 0 && j > 0) {
                    prefixY[i][j] -= prefixY[i-1][j-1];
                }
                
                if (prefixX[i][j] > 0 && prefixX[i][j] == prefixY[i][j]) {
                    count++;
                }
            }
        }
        
        return count;
    }
};

File: Dsa-day244.py
class Solution:
    def val(self, x):
        if x == '.': return 0
        if x == 'X': return 1
        return -1
    def numberOfSubmatrices(self, grid: List[List[str]]) -> int:
        n, m = len(grid), len(grid[0])
        s = set()
        vals = [[(0, False)] * m for _ in range(n)]        
        vals[0][0] = (self.val(grid[0][0]), self.val(grid[0][0]) == 1)
        res = 0
        for i in range(1, m):
            vals[0][i] = (vals[0][i-1][0] + self.val(grid[0][i]), vals[0][i-1][1] or self.val(grid[0][i]) == 1)
        for j in range(1, n):
            vals[j][0] = (vals[j-1][0][0] + self.val(grid[j][0]), vals[j-1][0][1] or self.val(grid[j][0]) == 1)
        for i in range(1, n):
            for j in range(1, m):
                vals[i][j] = (vals[i-1][j][0] + vals[i][j-1][0] - vals[i-1][j-1][0] + self.val(grid[i][j]), 
                vals[i-1][j][1] or vals[i][j-1][1] or (self.val(grid[i][j]) == 1))
        for i in range(n):
            for j in range(m):
                if vals[i][j][0] == 0 and vals[i][j][1] == True:
                    res += 1
        return res